### PR TITLE
WIP: Waku-rln-relay: Message validation and double signaling detection

### PR DESCRIPTION
**Pull request by staheri14**
Addresses 3rd and 4th items of https://github.com/status-im/nim-waku/issues/714
In the current implementation, double signaling can be detected using messages nullifiers, however, the recovery of the spammer's SK is not supported because the rln Rust lib lacks the secret reconstruction. Note that this does not harm the routing logic, because spam messages will be still identified and discarded (it is just not possible to slash the spammers).
  
In the next PRs 
- I will enable rln-relay compilation by default (https://github.com/status-im/nim-waku/issues/741) and will address the CI bug related to JSON-rpc API (as explained in this issue https://github.com/status-im/nim-waku/issues/750).
- I will update specs
- The final step would be to integrate the waku-rln-relay into chat2 and make it ready for dogfooding.

### CODE REVIEW

**Comment by staheri14** [comment]
Based on my investigation, the failure of test2 on `jenkins/prs/windows` is irrelevant to my PR. You can see that the same error occurs on this test PR https://github.com/status-im/nim-waku/pull/772 which contains no effective change over the master branch. This issue needs to be dealt with in a separate PR.

**Comment by oskarth** [comment]
What about linux? It also seems to be failing. Is this also related to the Websockets PR or something else?

File: waku/v2/node/wakunode2.nim [hunk 1]
```diff
@@ -440,10 +440,19 @@ when defined(rln):
...
-        if node.wakuRlnRelay.rlnInstance.proofVerify(msg.value().payload, msg.value().proof):
-          return ValidationResult.Accept
-      return ValidationResult.Reject
+        let 
+          wakumessage = msg.value()
+          # validate the message
+          validationRes = node.wakuRlnRelay.validateMessage(wakumessage)
+        case validationRes:
+          of Valid:
+            return ValidationResult.Accept
```
Comment (oskarth) [inline waku/v2/node/wakunode2.nim:449]: Maybe trace log here?

**Review by oskarth — APPROVED** [review]
Approving assuming CI problems definitely not part of this PR

**Comment by jm-clius** [comment]
> What about linux? It also seems to be failing. Is this also related to the Websockets PR or something else?

Linux failure in `deps` phase being tracked here: https://github.com/status-im/nim-waku/issues/738

File: waku/v2/node/wakunode2.nim [hunk 2]
```diff
@@ -440,10 +440,19 @@ when defined(rln):
...
+          of Invalid:
+            info "message validity could not be verified, discarding:", wakumessage=wakumessage
+            return ValidationResult.Reject
+          of Spam:
+            info "A spam message is found! yay! discarding:", wakumessage=wakumessage
```
Comment (jm-clius) [inline waku/v2/node/wakunode2.nim:454]: 👍 👍   Spam! Yay! 😄

**Review by jm-clius — APPROVED** [review]
LGTM!

**Review by rshiv — APPROVED** [review]
LGTM, please rebase to master , hopefully the windows CI should be passing.

File: waku/v2/node/wakunode2.nim (hunk 1 above)
Comment (staheri14) [reply waku/v2/node/wakunode2.nim:449]: Sure, done in https://github.com/status-im/nim-waku/pull/769/commits/754e62cfb564ececa3b2a555d348372844b22db9

**Comment by staheri14** [comment]
> LGTM, please rebase to master , hopefully the windows CI should be passing.

Thanks Reeshav! The windows issue is gone now

File: waku/v2/node/wakunode2.nim (hunk 2 above)
Comment (oskarth) [reply waku/v2/node/wakunode2.nim:455]: image

**Pull request merged**